File: tests/typesafety/python/money_pence_misuse.py
# expect: fails
# title: The swapped discount call, rejected by the checker
# misuse-of: money_pence
from typing import NewType

Pence = NewType("Pence", int)
Rate = NewType("Rate", float)


def apply_discount(total: Pence, rate: Rate) -> Pence:
    return Pence(round(total * (1 - rate)))


def checkout() -> Pence:
    return apply_discount(Rate(0.1), Pence(1250))

File: tests/typesafety/typescript/money_pence_misuse.ts
// expect: fails

declare const penceBrand: unique symbol;
declare const rateBrand: unique symbol;

type Pence = number & { readonly [penceBrand]: true };
type Rate = number & { readonly [rateBrand]: true };

function pence(n: number): Pence {
  return n as Pence;
}

function rate(n: number): Rate {
  return n as Rate;
}

function applyDiscount(total: Pence, discount: Rate): Pence {
  return pence(Math.round(total * (1 - discount)));
}

export const discounted = applyDiscount(rate(0.1), pence(1250));
